lib/shopping_cart.rb

```diff
@@ -0,0 +1,62 @@
+class ShoppingCart
+  attr_reader :name, :capacity, :products
+
+  def initialize(name, capacity)
+    @name = name
+    @capacity = capacity.to_i
+    @products = []
+  end
+
+  def add_product(product)
+    @products << product
+  end
+
+  def details
+    details_hash = {
+      name: @name,
+      capacity: capacity
+    }
+  end
+
+  def total_number_of_products
+    quantity_total = 0
+    products.each do |product|
+      quantity_total += product.quantity
+    end
+    
+    quantity_total
+  end
+
+  def is_full?
+   total_number_of_products > capacity 
+  end
+
+  def products_by_category(category)
+    filtered_products = products.find_all do |product|
+      product.category == category
+    end
+  end
+
+  def percentage_occupied
+    percent = ((total_number_of_products.to_f / capacity) * 100).round(2)
```

Comment:
same thing here.. and in the methods below... no need to save this into a variable if it's the laast line of your method anyway.